# IBM Capstone Project - 2nd Week - Car accident severity

# The 2nd Week

___

### Link to [the 1st Week](https://github.com/kolasdevpy/CapstoneProjectIBM/blob/master/Capstone%20Project%20%20-%201st%20Week%20-%20Car%20accident%20severity.ipynb) work

___

## Tasks

In this week, you will continue working on your capstone project. Please remember by the end of this week, you will need to submit the following:

<b>1)A full report consisting of all of the following components (15 marks):</b>

- <b>Introduction</b> where you discuss the business problem and who would be interested in this project. <b>(Done in [the 1st Week](https://github.com/kolasdevpy/CapstoneProjectIBM/blob/master/Capstone%20Project%20%20-%201st%20Week%20-%20Car%20accident%20severity.ipynb) work).</b>   

- Data where you <b>describe the data</b> that will be used to solve the problem and the source of the data. <b>(Done in [the 1st Week](https://github.com/kolasdevpy/CapstoneProjectIBM/blob/master/Capstone%20Project%20%20-%201st%20Week%20-%20Car%20accident%20severity.ipynb) work).</b>   

- <b>Methodology section which represents the main component of the report where you discuss and describe any exploratory data analysis</b> that you did, any inferential statistical testing that you performed, if any, and what machine learnings were used and why.  <b>The 2nd Week.</b>    

- <b>Results</b> section where you discuss the results. <b>The 2nd Week.</b>   

- <b>Discussion</b> section where you discuss any observations you noted and any recommendations you can make based on the results. <b>The 2nd Week.</b>   

- <b>Conclusion</b> section where you conclude the report. <b>The 2nd Week.</b>   
   
   
<b>2) A link to your <b>Notebook on your Github</b> repository pushed showing your code. (15 marks)</b> <b>The 2nd Week.</b>

<b>3) Your choice of a <b>presentation</b> or <b>blogpost</b>. (10 marks)</b> <b>The 2nd Week.</b>


___

# Remember the introduction

This data allows us to build a model for predicting whether a car crashes participants (car drivers or pedestrians) require increased amount of medical care or not. The data covers Seatle, WA. The primary focus of the model is to prioritize help for different points of the city for better balance of injury and help amount. The more injury - more help.

For example, the Seattle Police department has limited number of helicopters for the medical purposes and prosecution of criminals. In case of simultaneously happening crashes, the model could possibly help to choose the way of dealing with the problems, predicting the amount of damage in all cases.


# Reading data

We made a complete description, pre-processing and evaluation of the data in [the 1st week of the project](https://github.com/kolasdevpy/CapstoneProjectIBM/blob/master/Capstone%20Project%20%20-%201st%20Week%20-%20Car%20accident%20severity.ipynb).

In [1]:
import pandas as pd
import numpy as np

import os
path = os.path.expanduser("~/Documents/useful_df_car_accident_severity.csv")
df = pd.read_csv (path)

df.head(2)

Unnamed: 0           X          Y  ADDRTYPE  INTKEY  SEVERITYCODE  \
0           0 -122.323148  47.703140         1   37475             2   
1           1 -122.347294  47.647172         0       0             1   

   COLLISIONTYPE  PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  ...  SPEEDING  \
0              1            2         0            0  ...         0   
1              4            2         0            0  ...         0   

   SEGLANEKEY  CROSSWALKKEY  HITPARKEDCAR  year  month  day  hour  minute  \
0           0             0             0  2013      3   27    14      54   
1           0             0             0  2006     12   20    18      55   

   weekday_name  
0             2  
1             2  

[2 rows x 29 columns]

In [2]:
df.dtypes

Unnamed: 0          int64
X                 float64
Y                 float64
ADDRTYPE            int64
INTKEY              int64
SEVERITYCODE        int64
COLLISIONTYPE       int64
PERSONCOUNT         int64
PEDCOUNT            int64
PEDCYLCOUNT         int64
VEHCOUNT            int64
JUNCTIONTYPE        int64
SDOT_COLCODE        int64
INATTENTIONIND      int64
UNDERINFL           int64
WEATHER             int64
ROADCOND            int64
LIGHTCOND           int64
PEDROWNOTGRNT       int64
SPEEDING            int64
SEGLANEKEY          int64
CROSSWALKKEY        int64
HITPARKEDCAR        int64
year                int64
month               int64
day                 int64
hour                int64
minute              int64
weekday_name        int64
dtype: object

In [3]:
df.shape

(189339, 29)

In [4]:
X = df[['X', 'Y',
        'ADDRTYPE', 'INTKEY',
        'COLLISIONTYPE', 'PERSONCOUNT',
        'PEDCOUNT', 'PEDCYLCOUNT', 'VEHCOUNT',
        'JUNCTIONTYPE', 'SDOT_COLCODE', 'INATTENTIONIND', 
        'UNDERINFL', 'WEATHER', 'ROADCOND', 'LIGHTCOND',
        'PEDROWNOTGRNT', 'SPEEDING',
        'SEGLANEKEY', 'CROSSWALKKEY', 'HITPARKEDCAR',
        'year', 'month', 'day', 'hour', 'minute', 'weekday_name']]

y = df['SEVERITYCODE']

# Methodology section

### To do this, we need to build a model that can determine the need for a medical helicopter in an accident. It's a Discrete value.

### Therefore we have a Classification problem.

### Now we will try to understand what data we can get as quickly as possible or automatically from services, soft or witness. Next, we group their by sources.

> 'SEVERITYCODE' - target
   
   
> <b>1. Police get data from witness</b>   
> 'X'     
> 'Y'    
> 'COLLISIONTYPE'    
> 'PERSONCOUNT'    
> 'PEDCOUNT'    
> 'PEDCYLCOUNT'    
> 'VEHCOUNT'    
> 'ADDRTYPE'    
> 'HITPARKEDCAR'    


> <b>2. Police can get data automatically by post-processing coordinates 'X' and 'Y'</b>   
> 'INTKEY'    
> 'JUNCTIONTYPE'    
> 'SEGLANEKEY'    
> 'CROSSWALKKEY'    

> <b>3. Police can get data automatically by services</b>   
> 'WEATHER'    
> 'ROADCOND'    
> 'LIGHTCOND'  
> 'year'    
> 'month'    
> 'day'    
> 'hour'    
> 'minute'    
> 'weekday_name'    

> <b>4. Police can get data automatically by post-processing</b>   
> 'SDOT_COLCODE'   
> 'INATTENTIONIND'    
> 'UNDERINFL'   
> 'PEDROWNOTGRNT'
> 'SPEEDING'


### Obviously, that the information in Section 4 may be collected for a long time after the accident. Which makes us to understand  it is not relevant information, so we should drop these features.

> 'SEVERITYCODE' - target
   
   
> <b>1. Police get data from witness</b>   
> 'X'     
> 'Y'    
> 'COLLISIONTYPE'    
> 'PERSONCOUNT'    
> 'PEDCOUNT'    
> 'PEDCYLCOUNT'    
> 'VEHCOUNT'    
> 'ADDRTYPE'    
> 'HITPARKEDCAR'    


> <b>2. Police can get data automatically by post-processing coordinates 'X' and 'Y'</b>   
> 'INTKEY'    
> 'JUNCTIONTYPE'    
> 'SEGLANEKEY'    
> 'CROSSWALKKEY'    

> <b>3. Police can get data automatically by services</b>   
> 'WEATHER'    
> 'ROADCOND'    
> 'LIGHTCOND'  
> 'year'    
> 'month'    
> 'day'    
> 'hour'    
> 'minute'    
> 'weekday_name'

### Modeling and Evaluations

#### DecisionTreeClassifier

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

X =  df[['X', 'Y', 'COLLISIONTYPE', 'PERSONCOUNT', 'PEDCOUNT', 'PEDCYLCOUNT', 'VEHCOUNT', 'ADDRTYPE', 'HITPARKEDCAR', # 1 Section
         'INTKEY', 'JUNCTIONTYPE', 'SEGLANEKEY', 'CROSSWALKKEY',                                                      # 2 Section
         'WEATHER', 'ROADCOND', 'LIGHTCOND', 'year', 'month', 'day', 'hour', 'minute', 'weekday_name']]               # 3 Section

y = df['SEVERITYCODE']

X_trainset, X_testset, y_trainset, y_testset = train_test_split(X, y, test_size=0.1, random_state=3)
Tree = DecisionTreeClassifier(criterion="entropy")
Tree.fit(X_trainset,y_trainset)            
predTree = Tree.predict(X_testset)
accuracy = metrics.accuracy_score(y_testset, predTree)



print(f"DecisionTrees's Accuracy:")
print(f"{accuracy}")

DecisionTrees's Accuracy:
0.6837963451991127


#### LGBMClassifier

In [6]:
import lightgbm as lgb

X =  df[['X', 'Y', 'COLLISIONTYPE', 'PERSONCOUNT', 'PEDCOUNT', 'PEDCYLCOUNT', 'VEHCOUNT', 'ADDRTYPE', 'HITPARKEDCAR', # 1 Section
         'INTKEY', 'JUNCTIONTYPE', 'SEGLANEKEY', 'CROSSWALKKEY',                                                      # 2 Section
         'WEATHER', 'ROADCOND', 'LIGHTCOND', 'year', 'month', 'day', 'hour', 'minute', 'weekday_name']]               # 3 Section

y = df['SEVERITYCODE']

X_trainset, X_testset, y_trainset, y_testset = train_test_split(X, y, test_size=0.1, random_state=3)
lgbm = lgb.LGBMClassifier()
lgbm.fit(X_trainset,y_trainset)
predTree = lgbm.predict(X_testset)
accuracy = metrics.accuracy_score(y_testset, predTree)


print(f"LGBMClassifier Accuracy:")
print(f"{accuracy}")

/Users/artyomkolas/opt/anaconda3/envs/ibm/lib/python3.8/site-packages/lightgbm/__init__.py:42: UserWarning: Starting from version 2.2.1, the library file in distribution wheels for macOS is built by the Apple Clang (Xcode_8.3.3) compiler.
This means that in case of installing LightGBM from PyPI via the ``pip install lightgbm`` command, you don't need to install the gcc compiler anymore.
Instead of that, you need to install the OpenMP library, which is required for running LightGBM on the system with the Apple Clang compiler.
You can install the OpenMP library by the following command: ``brew install libomp``.
  warnings.warn("Starting from version 2.2.1, the library file in distribution wheels for macOS "


LGBMClassifier Accuracy:
0.7583183690715115


# Results

DecisionTrees's Accuracy:
0.6828456744480829

LGBMClassifier Accuracy:
0.7580542938628921

We can see that LGBMClassifier performs better.
I propose to dwell on this solution.

Our model is able to predict whether a medical assistance is needed or not with accuracy ~ 76%. It's good.

# Discussion

### We can pay attention to the order in which information about the incident is received.

### 1) If a witness to the incident exists at the time of the incident, the dispatcher can obtain information in section 1 and automatically and quickly obtain information from sections 2 and 3.

Let's consider some options:   
A police dispatcher can spend 1-5 minutes collecting information.    
The information may be complete or incomplete.    
The information may be true or false.    
The incident may be life threatening or not.   

#### In this case, we have a lot of conventions. But a witness can tell us about the wounded and the need for medical assistance. We can compare the report to the police with the decision from the model and draw conclusions about the credibility of the information.    
#### Or if there is not enough information about the victims, can help to police to make a decision.   
#### The model can help is to understand immediately in the first seconds of the call (by coordinates) whether medical assistance is needed or not.

### 2) If a witness to the incident does not exist at the time of the accident (or appears after a long time), then the dispatcher will not receive information at all, or the lost time may cost the lives of the victims.

Let's consider some options   
The witness does not appear at all, and the victims do not need medical assistance.
The witness does not appear at all, and the victims need medical attention but do not receive it.
An witness appears, but after a while the victims will receive help (if it is necessary).
An witness appears, but after a while the victims do not have time to get help (if it is necessary)


#### In this case, we have no information at all at the time of the accident, moreover if the victims need medical assistance, they may not receive it.   
#### What can we do about it?
1) we could get information from the navigator, for example, about an abnormal decrease in speed. After that automatically transfer coordinates to the dispatcher.   
2) we could receive information from the deployment of airbags, sensors of the integrity of the bumper or car body, impact on the body, and then automatically transmit the coordinates to the dispatcher.

Based on this we could get such data from software and services and build a model.

   
> <b>1. Police get data from witness</b>   
> 'X'     
> 'Y'    


> <b>2. Police can get data automatically by post-processing coordinates 'X' and 'Y'</b>   
> 'INTKEY'    
> 'JUNCTIONTYPE'    
> 'SEGLANEKEY'    
> 'CROSSWALKKEY'    

> <b>3. Police can get data automatically by services</b>   
> 'WEATHER'    
> 'ROADCOND'    
> 'LIGHTCOND'  
> 'year'    
> 'month'    
> 'day'    
> 'hour'    
> 'minute'    
> 'weekday_name'

Such methods are able to help to reduce the number of victims where every minute is important.

In this case, our model can help to inform the dispatcher about the severity of the accident for less than a second and send a medical helicopter before dispather can contact the victims.

In [7]:
import lightgbm as lgb


X =  df[['X', 'Y',                                                                                                    # 1 Section
         'INTKEY', 'JUNCTIONTYPE', 'SEGLANEKEY', 'CROSSWALKKEY',                                                      # 2 Section
         'WEATHER', 'ROADCOND', 'LIGHTCOND', 'year', 'month', 'day', 'hour', 'minute', 'weekday_name']]               # 3 Section

y = df['SEVERITYCODE']

X_trainset, X_testset, y_trainset, y_testset = train_test_split(X, y, test_size=0.1, random_state=3)
lgbm = lgb.LGBMClassifier()
lgbm.fit(X_trainset,y_trainset)
predTree = lgbm.predict(X_testset)
accuracy = metrics.accuracy_score(y_testset, predTree)


print(f"LGBMClassifier Accuracy:")
print(f"{accuracy}")

LGBMClassifier Accuracy:
0.7228794760747861


# Conclusion

### 1) Our model can help the police to define the need of medical assistance in case of simultaneously happening crashes with 76 % accuracy.
### 2) We can compare police report with the decision from the model and draw conclusions about the credibility of the information with 76 % accuracy.
### 3) If there is not enough information about the accident the model can help the police to make a decision about medical assistance with accuracy  ~ 72 % (by coordinates).
### 4) The model can help us to understand in the first seconds of the call   (by coordinates) whether medical assistance is needed or not with accuracy  ~ 72 %.
### 5) If cars are equipped with accident warning systems for the police our model can help to decide whether to provide medical assistance or not with accuracy  ~ 72 % (by coordinates).


# That's all, folks

Created by [Artyom Kolas ](https://www.linkedin.com/in/artyomkolas/)